hack/generator/pkg/codegen/pipeline_remove_embedded_resources.go

```diff
@@ -0,0 +1,681 @@
+		_, ok := astmodel.AsResourceType(def.Type())
+		return ok
+	})
+
+	var errs []error
+	result := make(map[astmodel.TypeName]astmodel.TypeNameSet)
+
+	// Identify sub-resources and their "properties", associate them with parent resource
+	// Look through parent resource for subresource properties
+	for _, def := range resources {
+		resource, ok := astmodel.AsResourceType(def.Type())
+		if !ok {
+			// Shouldn't be possible to get here
+			panic(fmt.Sprintf("resource was somehow not a resource: %q", def.Name()))
+		}
+
+		if resource.Owner() != nil {
+			owner := *resource.Owner()
+			if result[owner] == nil {
+				result[owner] = make(astmodel.TypeNameSet)
+			}
+
+			specPropertiesTypeName, statusPropertiesTypeName, err := extractSpecAndStatusPropertiesTypeOrNil(types, resource)
+			if err != nil {
+				errs = append(errs, errors.Wrapf(err, "couldn't extract spec/status properties from %q", def.Name()))
+				continue
+			}
+			if specPropertiesTypeName != nil {
+				result[owner] = result[owner].Add(*specPropertiesTypeName)
+			}
+			if statusPropertiesTypeName != nil {
+				result[owner] = result[owner].Add(*statusPropertiesTypeName)
+			}
+		}
+	}
+
+	if len(errs) > 0 {
+		return nil, kerrors.NewAggregate(errs)
+	}
+
+	return result, nil
+}
+
+func extractSpecAndStatusPropertiesTypeOrNil(types astmodel.Types, resource *astmodel.ResourceType) (*astmodel.TypeName, *astmodel.TypeName, error) {
+	specPropertiesTypeName, err := extractSpecPropertiesTypeNameOrNil(types, resource)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't extract spec properties")
+	}
+
+	statusPropertiesTypeName, err := extractStatusPropertiesTypeNameOrNil(types, resource)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't extract status properties")
+	}
+
+	return specPropertiesTypeName, statusPropertiesTypeName, nil
+}
+
+// findAllResourcePropertiesTypes finds the "Properties" type for each resource. The result is a astmodel.TypeNameSet containing
+// each resources "Properties" type.
+func findAllResourcePropertiesTypes(types astmodel.Types) (astmodel.TypeNameSet, error) {
+	resources := types.Where(func(def astmodel.TypeDefinition) bool {
+		_, ok := astmodel.AsResourceType(def.Type())
+		return ok
+	})
+
+	var errs []error
+	result := make(astmodel.TypeNameSet)
+
+	// Identify sub-resources and their "properties", associate them with parent resource
+	// Look through parent resource for subresource properties
+	for _, def := range resources {
+		resource, ok := astmodel.AsResourceType(def.Type())
+		if !ok {
+			// Shouldn't be possible to get here
+			panic(fmt.Sprintf("resource was somehow not a resource: %q", def.Name()))
+		}
+
+		specPropertiesTypeName, statusPropertiesTypeName, err := extractSpecAndStatusPropertiesTypeOrNil(types, resource)
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "couldn't extract spec/status properties from %q", def.Name()))
+			continue
+		}
```

Comment:
`extractSpecAndStatusPropertiesTypeOrNil` is actually used 2 places and so is reducing _some_ duplication, whereas the individual `extractSpec` and `extractStatus` ones are only used in the one place and basically add no value

How about I rename `extractSpecAndStatusPropertiesTypeOrNil` to `tryResolveSpecStatusTypes` and remove the `extractSpec`/`extractStatus` layer?

Reply:
Doing the above leaves something like this:
```
func tryResolveSpecStatusTypes(types astmodel.Types, resource *astmodel.ResourceType) (*astmodel.TypeName, *astmodel.TypeName, error) {
	specName, ok := astmodel.AsTypeName(resource.SpecType())
	if !ok {
		return nil, nil, errors.Errorf("resource spec was not a TypeName")
	}

	specPropertiesTypeName, err := followTypeNameExtractPropertiesType(types, specName)
	if err != nil {
		return nil, nil, errors.Wrap(err, "couldn't extract spec properties")
	}

	var statusPropertiesTypeName *astmodel.TypeName
	statusName, ok := astmodel.AsTypeName(resource.StatusType())
	if ok {
		statusPropertiesTypeName, err = followTypeNameExtractPropertiesType(types, statusName)
		if err != nil {
			return nil, nil, errors.Wrap(err, "couldn't extract status properties")
		}
	}

	return specPropertiesTypeName, statusPropertiesTypeName, nil
}
```

Which feels meaty enough to me to avoid the axe?

Reply:
> Are the names long to avoid collisions with functions from the other pipeline stages in the same package?

No I think I'm just guilty of writing names that are too close to the implementation of the function. I like the idea of pushing some of this to its own sub-directory though, will look into that.

Reply:
Ah, the issue with moving the whole stage into a subdirectory is that the stage itself depends on `PipelineStage` which is currently defined in `codgen`, so there'd be a circular reference if I moved it.

I can move the logic though and just leave the stage (which actually feels more in-keeping with what we've been doing anyway...)